### PR TITLE
Tsallis Entropy Shift and more

### LINKED ISSUE #12: Allow user to pass their own ax for a shift graph

### PR DESCRIPTION
**Pull request by cvanoort**
Sorry it's taken me so long to get around to this!

I have an initial implementation of a Tsallis Entropy Shift, it follows the same interface as the JSD shift but uses Tsallis entropy instead of Shannon's entropy and thus allows for tuning of the entropy order.

I also cleaned up a few things, stale imports, dead code, and inconsistent code style.

I added a few integration tests that should help with diagnosing bugs during development. Unfortunately they aren't fully automated at the moment, the test needs to be done mostly by eye to ensure that the figures look correct, but they've already helped me to catch a few bugs, so I think that they will be useful.

Finally, I added the ability to pass an axis object to Shift.get_shift_plot, so this should close #12 I think.

I'll try to make the next PR smaller so that it's easier to review!

### CODE REVIEW

**Comment by ryanjgallagher** [comment]
Wow, thank you for all your work on this. I'll need a little bit of time to look through this, so apologies in advance for the delay merging it.

Sorry, I didn't realize you were working on the Tsallis entropy, I wouldn't have made so many big structural changes upstream. It looks like you were able to resolve the conflicts on your end, but I think we might have two implementations of Tsallis entropy now (I did one just earlier this week). I moved all the entropy helper functions to [entropy.py](https://github.com/ryanjgallagher/shifterator/blob/master/shifterator/entropy.py), and then implemented Tsallis entropy through the `alpha` parameter for `EntropyShift` and `JSDivergenceShift`. 

What's the best way to go forward with that? I like wrapping it in the entropy shift because it's more compact, but I probably haven't tested mine as much around the edge cases as yours.

I'm working on docs that can be hosted on Read the Docs, and one of my last major things was to make the docstrings for the shift classes better. I'll hold off on that until we get this merged so that we don't have to deal with more unnecessary conflicts.

**Comment by cvanoort** [comment]
Totally understandable, take your time.

We can compare the two implementations and keep the parts that look the best from each, there's definitely no reason to have both. I agree that folding the Tsallis entropy into the EntroyShift and JSDShift is probably better in the long run, I only made a separate shift class to allow for direct side-by-side comparison of the TsallisDivergenceShift and the JSDShift. There are some subtle differences at the moment, which we might want to check out before we finalize things here.

I noticed that you were working on the doc strings, what are you planning on using to build the docs? Sphinx, or something else?

**Comment by ryanjgallagher** [comment]
If you have time, would you be able to look into the differences between the implementations of the entropy-based shifts?

Yes, I'm using Sphinx to build the docs.

**Comment by cvanoort** [comment]
Alright, I think I managed to mush together the two implementations and clean things up a little. One quick question, shouldn't the exponent in the Tsallis part be alpha, rather than [alpha - 1](https://github.com/ryanjgallagher/shifterator/blob/95a58c941522454fcf911e429fc920bc38540991/shifterator/entropy.py#L96)?

**Comment by ryanjgallagher** [comment]
Great! I should be able to take a closer look at this next week hopefully.

The word shift graphs write every measure as a weighted average like \sum p_i * s_i, where p_i is the relative frequency and s_i is the word's score. For the Tsallis entropy, the overall formula has p_i^alpha, but we pull out one p_i so that we can write it as p_i * p_i^(alpha - 1) = p_i * s_i. Does that make sense?

File: shifterator/__init__.py [hunk 1]
```diff
@@ -1 +1,5 @@
+from .shifterator import Shift
+from .shifts import (EntropyShift, JSDivergenceShift, KLDivergenceShift,
```
Comment (ryanjgallagher) [inline shifterator/__init__.py]: Add WeightedAvgShift here too

File: shifterator/helper.py [hunk 2]
```diff
@@ -72,38 +72,38 @@ def get_score_dictionary(scores):
...
-        elif 'SocialSent' in lexicon:
+        elif "SocialSent" in lexicon:
             lexicon_ref = 0
-        elif 'NRC' in lexicon:
+        elif "NRC" in lexicon:
             lexicon_ref = 0.5
     except FileNotFoundError:
-        raise FileNotFoundError('Lexicon does not exit in Shifterator: {}'.format(scores))
+        raise FileNotFoundError(
+            "Lexicon does not exit in Shifterator: {}".format(scores)
```
Comment (ryanjgallagher) [inline shifterator/helper.py]: This is minor: if you could fix the typo here ("exists") that would be helpful. I know the typo wasn't introduced by this pull, just noticed as I was going through

File: shifterator/helper.py [hunk 3]
```diff
@@ -115,11 +115,14 @@ def get_missing_scores(type2score_1, type2score_2):
     type2score_1, type2score_2: dict
         Keys are types and values are scores
 
-    Output
+    Returns
     ------
```
Comment (ryanjgallagher) [inline shifterator/helper.py]: Just add one more "-" so it goes all the way under "Returns". Think it might mess up automated documentation otherwise

File: shifterator/plotting.py [hunk 4]
```diff
@@ -1,132 +1,83 @@
...
+            "pos_s": "#FECC5D",
+            "pos_s_neg_p": "#FDFFD2",
+            "pos_s_pos_p": "#FFFF80",
+            "pos_total": "#FFFF80",
+            "total": "#707070",
+        },
+        "serif": False,
+        "show_score_diffs": show_score_diffs,
+        "show_total": True,
+        "symbols": {
```
Comment (ryanjgallagher) [inline shifterator/plotting.py:48]: Note here that in resolving merge conflicts, a little bit got lost in translation. 

I think it should be:
```
'pos_total': plot_params['system_names'][1],	
'neg_total': plot_params['system_names'][0],	
'all_pos_pos': plot_params['system_names'][1],	
'all_pos_neg': plot_params['system_names'][0],
```

Does this happen to resolve #19 if you make this change?

File: shifterator/plotting.py [hunk 5]
```diff
@@ -1,132 +1,83 @@
...
+        "dpi": 200,
+        "height": 15,
+        "invisible_spines": [],
+        "label_fontsize": 13,
+        "missing_symbol": "*",
+        "pos_cumulative_inset": [0.19, 0.12, 0.175, 0.175],
+        "pos_text_size_inset": [0.81, 0.12, 0.08, 0.08],
+        "remove_xticks": False,
+        "remove_yticks": False,
+        "score_colors": {
```
Comment (ryanjgallagher) [inline shifterator/plotting.py:32]: These changes that got lost between commits should be updated:

```
'pos_total': '#FECC5D',	
'neg_total': '#9E75B7',	
'all_pos_pos': '#FECC5D',	
'all_pos_neg': '#9E75B7',
```

File: shifterator/plotting.py [hunk 6]
```diff
@@ -280,32 +249,76 @@ def plot_contributions(ax, top_n, bar_dims, bar_colors, plot_params):
         `detailed`, and `bar_linewidth` are used
     """
     # Set plotting params
-    ys = range(1, top_n + 1)
-    alpha = plot_params['alpha_fade']
-    width = plot_params['bar_width']
-    linewidth = plot_params['bar_linewidth']
-    edgecolor = ['black'] * top_n # hack b/c matplotlib has a bug
-    if plot_params['detailed']:
+    bar_count = min(top_n, len(bar_dims["total_heights"]))
```
Comment (ryanjgallagher) [inline shifterator/plotting.py:254]: In what situation is `bar_count` needed? Is it when there are less than `top_n` words? Does everything output correctly if there are less than `top_n` words? I haven't tried that situation before I think.

**Comment by ryanjgallagher** [comment]
This all looks great @cvanoort. I left some comments for mostly small things to change. One of them may be related to #19 and help address it. If it doesn't, just let me know and I can fix that in a separate pull request.

Once you make those updates, I should be able to merge.

File: shifterator/plotting.py (hunk 4 above)
Comment (ryanjgallagher) [reply shifterator/plotting.py:48]: Actually I'm looking locally and I was doing something a bit more complicated. 

```python
if diff >= 0:
    pos_sys_name = plot_params['system_names'][1]
    neg_sys_name = plot_params['system_names'][0]
else:
    pos_sys_name = plot_params['system_names'][0]
    neg_sys_name = plot_params['system_names'][1]
    plot_params['symbols'] = {
        'pos_s_pos_p': u'+\u2191',
        'pos_s_neg_p': u'+\u2193',
        'neg_s_pos_p': u'-\u2191',
        'neg_s_neg_p': u'-\u2193',
        'pos_s': u'\u25B3',
        'neg_s': u'\u25BD',
        'pos_total': pos_sys_name,
        'neg_total': neg_sys_name,
        'all_pos_pos': plot_params['system_names'][1],
        'all_pos_neg': plot_params['system_names'][0],
        'total': r'$\Sigma$'}
```

`diff` had to be added to the `get_plot_params` signature, and it was passed `self.diff` when it's called.

I think I hadn't decided if this was the right fix, which is why I hadn't pushed it yet. Let me know your thoughts on this.

File: shifterator/plotting.py (hunk 6 above)
Comment (cvanoort) [reply shifterator/plotting.py:254]: Yup, this was just to make things look right when there weren't enough words. It should almost never get encountered in the wild, but some of my tiny test cases didn't look right without it.

File: shifterator/plotting.py (hunk 4 above)
Comment (cvanoort) [reply shifterator/plotting.py:48]: #19 is gone with these changes!

Comment (cvanoort) [reply shifterator/plotting.py:48]: As for the second snippet, I think it's the same as what I have for all of my test cases, so if it's more correct or nicer to work with for some of the cases that you have then I don't think there's much of a reason to not have it. I'll roll it into the PR.
